### PR TITLE
Allow configuration of GSettings proxy

### PR DESCRIPTION
**Pull request by GabrielNagy**
Similar to APT and environment proxy, GSettings configuration is done by writing to a file, with the addition of executing `glib-compile-schemas` afterwards in order to make GSettings aware of the changes.

Additionally, there's more parsing involved here due to the specific format required the override files, and the `org.gnome.system.proxy` GSchema.

Because of the need to execute a shell command, more intricacies are involved, such as executing the command if we are asked to remove all settings, and handling command execution failure by reverting to the previous version of the GSchema override file.

The ubuntu-proxy-manager service is designed to run on Ubuntu machines, but there may be valid cases where users would like to run it on other flavors that aren't necessarily GNOME-based - so if we can't find the `glib-compile-schemas` command, we just log a message and cleanly exit.

DEENG-561

This PR contains a couple more fixes for `no_proxy` and adjustments to support the `auto` proxy option (which is only supported by GSettings), not necessarily related to this PR but spotted while working on it. Thus this is best reviewed commit-by-commit.

### CODE REVIEW

File: internal/proxy/gsettings.go
Comment (denisonbarbosa) [inline internal/proxy/gsettings.go:115]: If we need a restore point in case we need to restore the schemas, maybe it's better to, instead of overwriting the file in the first place, rename the previous one to `config.old` or `config.prev` and then restore it to `config` if we fail to apply the new one, wdyt? It will take some refactoring, but it will take less writing operations in the end.

Ps. Maybe we can go for a bool argument in `safeWriteFile` to determine if we should backup the old file or not, wdyt?

File: internal/proxy/proxy.go
Comment (denisonbarbosa) [inline internal/proxy/proxy.go:88]: That change from before feels so good, I'm glad we went this way... Well done!

File: internal/proxy/testdata/TestApply/golden/all_options_set_and_equal,_all_proxy_is_set/usr/share/glib-2.0/schemas/99_ubuntu-proxy-manager.gschema.override
Comment (denisonbarbosa) [inline internal/proxy/testdata/TestApply/golden/all_options_set_and_equal,_all_proxy_is_set/usr/share/glib-2.0/schemas/99_ubuntu-proxy-manager.gschema.override:19]: Is this on purpose?

**Review by denisonbarbosa — CHANGES_REQUESTED** [review]
A couple of suggestions and honest questions, but everything is great. Nice work as always!

File: internal/proxy/env.go
Comment (denisonbarbosa) [reply internal/proxy/env.go:22]: In the end, it's up to you. I feel like, when it comes to user feedback, it can never be too much. I agree that logging it 5 times is a bit annoying, but for this kind of errors I feel like it's important to be explicit about it, no? Even though we already document it, it doesn't hurt to log it back to the users should them try to set an unsupported protocol (specially since they differ depending on the proxy setting).
Nevertheless, the final decision is yours.

File: internal/proxy/gsettings.go
Comment (didrocks) [inline internal/proxy/gsettings.go:121]: The error you are wrapping is the error of `moveBack()`, not the one of `safeWriteFile()` or `runGlibCompileSchemas()`. Also, I find it weird that errors.Join() has a side-effect for restoring.

I would just have them in 3 lines:
- `errors.Join()` with the current error
- `moveBack()`
- if error on moveBack, `errors.Join()` it

File: internal/proxy/gsettings.go
Comment (didrocks) [inline internal/proxy/gsettings.go:85]: isn’t that `glibSchemasPath`?

File: internal/proxy/proxy.go
Comment (didrocks) [inline internal/proxy/proxy.go:67]: I would assign a variable `glibSchemasPath` just above, directly assign it to it in `glibSchemasPath` and `filepath.Join(glibSchemasPath, gschemaOverrideFile)` to get `gsettingsConfigPath`.

That way, you remove a potential divergence in code using them.

File: internal/proxy/proxy.go
Comment (didrocks) [inline internal/proxy/proxy.go:142]: > // If the file doesn't exist, no error is returned.
I find that behaviour weird though? I would expect the original file not existing at `path` to return an error.

File: internal/proxy/proxy_test.go
Comment (didrocks) [inline internal/proxy/proxy_test.go:47]: `glibMockNotRun` is rather a "want" value, no?

**Review by didrocks — CHANGES_REQUESTED** [review]
I haven’t looked at the golden files updates yet, but all the rest looks good to me, with some comments to simplify things :) Very well done overall (and on the test implementation of it too)!

File: internal/proxy/proxy_test.go
Comment (didrocks) [reply internal/proxy/proxy_test.go:47]: You are using it in:
```go
			if !tc.glibMockError && !tc.glibMockNotRun && !tc.missingGlibExecutable {
				err := os.Remove(filepath.Join(root, glibCompileSchemasRunFile))
				require.NoError(t, err, "glib-compile-schemas mock was not executed")
			}
```

Which is the assertion part (or maybe you should check explicitely that the file is not create if the boolean is true), but this is a little bit like a `wantErr`, no?

Comment (didrocks) [reply internal/proxy/proxy_test.go:47]: Maybe we should have one parameter `wantGlibMockRun` and then, keep the 2 others as fixtures input, wdyt?

File: internal/proxy/gsettings.go
Comment (didrocks) [reply internal/proxy/gsettings.go:174]: I would then still uninline it, to make it clear, but I see your point, tell me what you do prefer :)

File: internal/proxy/gsettings.go
Comment (didrocks) [reply internal/proxy/gsettings.go]: I will need to get used to join a nil error, but reading the standard library, that works :)

File: internal/proxy/testdata/TestApply/golden/some_options_set,_configuration_parent_directories_are_created/etc/environment.d/99ubuntu-proxy-manager.conf
Comment (didrocks) [inline internal/proxy/testdata/TestApply/golden/some_options_set,_configuration_parent_directories_are_created/etc/environment.d/99ubuntu-proxy-manager.conf:1]: Do you mind checking this is expected?

**Review by didrocks — APPROVED** [review]
Very well done! Approved once you have double checked one of the golden file change.

File: internal/proxy/testdata/TestApply/golden/some_options_set,_configuration_parent_directories_are_created/etc/environment.d/99ubuntu-proxy-manager.conf
Comment (didrocks) [reply internal/proxy/testdata/TestApply/golden/some_options_set,_configuration_parent_directories_are_created/etc/environment.d/99ubuntu-proxy-manager.conf:1]: Ah, indeed, too may changes (as basically every line is changed) for this poor git :)

**Review by denisonbarbosa — APPROVED** [review]

**Pull request merged**
